Reaktoro/Core/Phases.cpp

```diff
@@ -1,350 +1,361 @@
+    auto specieswithouttags = species.withoutTags(excludetags);
+
+    Phase phase;
+    phase = phase.withName(phasename);
+    phase = phase.withStateOfMatter(stateofmatter);
+    phase = phase.withSpecies(specieswithouttags);
+    phase = phase.withActivityPropsFn(activity_model(specieswithouttags));
+    phase = phase.withIdealActivityPropsFn(ideal_activity_model(specieswithouttags));
+
+    return phase;
+}
+
+
+GenericPhasesGenerator::GenericPhasesGenerator()
+{}
+
+GenericPhasesGenerator::GenericPhasesGenerator(const StringList& species)
+: names(species)
+{}
+
+GenericPhasesGenerator::GenericPhasesGenerator(const Speciate& elements)
+: symbols(elements.symbols)
+{}
+
+GenericPhasesGenerator::~GenericPhasesGenerator()
+{}
+
+auto GenericPhasesGenerator::setStateOfMatter(StateOfMatter option) -> GenericPhasesGenerator&
+{
+    stateofmatter = option;
+    return *this;
+}
+
+auto GenericPhasesGenerator::setAggregateState(AggregateState option) -> GenericPhasesGenerator&
+{
+    aggregatestate = option;
+    return *this;
+}
+
+auto GenericPhasesGenerator::setActivityModel(const ActivityModel& model) -> GenericPhasesGenerator&
+{
+    activity_model = model;
+    return *this;
+}
+
+auto GenericPhasesGenerator::setIdealActivityModel(const ActivityModel& model) -> GenericPhasesGenerator&
+{
+    ideal_activity_model = model;
+    return *this;
+}
+
+auto GenericPhasesGenerator::set(StateOfMatter option) -> GenericPhasesGenerator&
+{
+    return setStateOfMatter(option);
+}
+
+auto GenericPhasesGenerator::set(AggregateState option) -> GenericPhasesGenerator&
+{
+    return setAggregateState(option);
+}
+
+auto GenericPhasesGenerator::set(const ActivityModel& model) -> GenericPhasesGenerator&
+{
+    return setActivityModel(model);
+}
+
+auto GenericPhasesGenerator::stateOfMatter() const -> StateOfMatter
+{
+    return stateofmatter;
+}
+
+auto GenericPhasesGenerator::aggregateState() const -> AggregateState
+{
+    return aggregatestate;
+}
+
+auto GenericPhasesGenerator::species() const -> const Strings&
+{
+    return names;
+}
+
+auto GenericPhasesGenerator::elements() const -> const Strings&
+{
+    return symbols;
+}
+
+auto GenericPhasesGenerator::activityModel() const -> const ActivityModel&
+{
+    return activity_model;
+}
+
+auto GenericPhasesGenerator::idealActivityModel() const -> const ActivityModel&
+{
+    return ideal_activity_model;
+}
+
+auto GenericPhasesGenerator::convert(const Database& db, const Strings& elements) const -> Vec<GenericPhase>
+{
+    error(aggregatestate == AggregateState::Undefined,
+        "GenericPhasesGenerator::convert requires an AggregateState value to be specified. "
+        "Use method GenericPhasesGenerator::set(AggregateState) to fix this.");
+
+    auto species = db.speciesWithAggregateState(aggregatestate);
+
+    species =
+        names.size() ? species.withNames(names) :
+        symbols.size() ? species.withElements(symbols) :
+            species.withElements(elements);
+
```

Comment:
We need here a similar logic to filter out the species with given tags. 

~~~cpp
if(excludetags().size())
    species = species.withoutTags(excludetags);
~~~